Collector/ProfilePlugin.php

```diff
@@ -77,15 +77,60 @@ public function handleRequest(RequestInterface $request, callable $next, callabl
             return $first($request);
         };
 
-        return $this->plugin->handleRequest($request, $wrappedNext, $wrappedFirst)->then(function (ResponseInterface $response) use ($profile) {
+        try {
+            $promise = $this->plugin->handleRequest($request, $wrappedNext, $wrappedFirst);
+        } catch (Exception $e) {
+            $profile->setFailed(true);
+            $profile->setResponse($this->formatter->formatException($e));
+            $this->collectRequestInformation($request, $stack);
+
+            throw $e;
+        }
+
+        return $promise->then(function (ResponseInterface $response) use ($profile, $request, $stack) {
             $profile->setResponse($this->formatter->formatResponse($response));
+            $this->collectRequestInformation($request, $stack);
 
             return $response;
-        }, function (Exception $exception) use ($profile) {
+        }, function (Exception $exception) use ($profile, $request, $stack) {
             $profile->setFailed(true);
```

Comment:
This exception can be thrown from anywhere too.

But, as some `Http\Promise\Promise` implementation does not catch all exceptions, we have to use both to catch exceptions in any case. See [`Http\Promise\FulfilledPromise`](https://github.com/php-http/promise/blob/master/src/FulfilledPromise.php) and [`Http\Client\Promise\HttpfulfilledPromise`](https://github.com/php-http/httplug/blob/master/src/Promise/HttpFulfilledPromise.php) for implementation differences.